relay-server/src/services/processor/nnswitch.rs

```diff
@@ -0,0 +1,428 @@
+//! Nintendo Switch crash reports processor related code.
+//!
+//! These functions are included only in the processing mode.
+
+use crate::envelope::{ContentType, EnvelopeError, Item, ItemType};
+use crate::services::processor::{ErrorGroup, ProcessingError};
+use crate::utils::{self, TypedEnvelope};
+use crate::Envelope;
+use bytes::{Buf, Bytes};
+use std::sync::OnceLock;
+use zstd::bulk::Decompressor as ZstdDecompressor;
+
+type Result<T> = std::result::Result<T, SwitchProcessingError>;
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum SwitchProcessingError {
+    #[error("invalid json")]
+    InvalidJson(#[source] serde_json::Error),
+    #[error("envelope parsing failed")]
+    EnvelopeParsing(#[from] EnvelopeError),
+    #[error("unexpected EOF, expected {expected:?}")]
+    UnexpectedEof { expected: String },
+    #[error("invalid {0:?} ({1:?})")]
+    InvalidValue(String, usize),
+    #[error("Zstandard error")]
+    Zstandard(#[source] std::io::Error),
+}
+
+/// Expands Nintendo Switch crash-reports.
+///
+/// If the envelope does NOT contain a `dying_message.dat` attachment, it doesn't do anything.
+/// If the attachment item is found and it matches the expected format, it parses the attachment
+/// and updates the envelope with the content. The processed attachment is then removed from the
+/// envelope.
+///
+/// The envelope may be dropped if it exceeds size limits after decompression. Particularly,
+/// this includes cases where a single attachment file exceeds the maximum file size. This is in
+/// line with the behavior of the envelope endpoint.
+///
+/// After this, [`crate::services::processor::EnvelopeProcessorService`] should be able to process
+/// the envelope the same way it processes any other envelopes.
+///
+/// Note: in case of an error, we don't fail but leave the envelope as is.
+pub fn expand(
+    managed_envelope: &mut TypedEnvelope<ErrorGroup>,
+) -> std::result::Result<(), ProcessingError> {
+    let envelope: &mut &mut crate::Envelope = &mut managed_envelope.envelope_mut();
+
+    if let Some(item) = envelope.take_item_by(is_dying_message) {
+        if let Err(e) = expand_dying_message(item.payload(), envelope) {
+            // If we fail to process the dying message, we need to add back the original attachment.
+            envelope.add_item(item);
+            return Err(ProcessingError::InvalidNintendoDyingMessage(e));
+        }
+    }
+
+    Ok(())
+}
+
+/// Magic number indicating the dying message file is encoded by sentry-switch SDK.
+const SENTRY_MAGIC: &[u8] = "sntr".as_bytes();
+
+/// The file name that Nintendo uses to in the events they forward.
+const DYING_MESSAGE_FILENAME: &str = "dying_message.dat";
+
+const MAX_DECOMPRESSED_SIZE: usize = 100_1024;
```

Comment:
Is that supposed to be `100 * 1024`?

Also small comment would be nice, what the limit is and why it exists.

Reply:
[MAX_DECOMPRESSED_SIZE](https://doc.rust-lang.org/rust-by-example/primitives/literals.html#literals-and-operators)

Reply:
> Also small comment would be nice, what the limit is and why it exists.

added a comment